minetest/minetest
# Use an sqlite database for the media cache

- Replaces the `cache_path/media/` folder with an sqlite database `cache_path/media_db.sqlite`.
  Most notable new features are cache cleaning and the possibility to store the files in the cache zlib-compressed.
- Mostly fixes #10598. This is just a very simple approach though. Also, by default nothing is deleted yet.
- How the PR works:
  * Instead of `path_cache/media/many_files`, there's now `path_cache/media_db.sqlite`.
  * The db contains together with the file data and sha1 digest also some metadata per media file: a cache-class and a data-format, and also a timestamp (in seconds since epoch) and a expiry counter (see further below):
    - The data-format says how the file data is encoded, ie. raw or zlib-compressed (only raw is used for now).
    - The cache-class (see enum `MediaCacheClass`) determines how long the file persists in the storage.
      * There are two cache-classes for now: `Legacy` (what we currently use) and `Unknown` (should be used if the server suggested a cache-class that we don't know).
      * Cache-class expiry times can be set by the user via settings (can also be "never" to never expire files).
  * The db is accessed twice on startup:
    - When the media announcement is received, all files are looked up via their sha1 digest, and their timestamp is updated.
    - When new media files are received, they are stored in ram. After all files were received, the db is updated with the new files, and some maintenance is done (see next point). This is done in a separate thread, so the game can start before it's done. (Note that updating all files at once is much faster than doing one transaction per file.)
  * The maintenance mentioned above includes cache expiry and vacuuming:
    - After a cache-class dependent time, a file expires (expire counter incremented, timestamp reset).
    - A file is removed after it expired twice.
    - After 1000 deleted files, the db is vacuumed. (There's a counter in a separate db table for db-wide metadata.) TODO: add a setting for this number
- roadmap: count this to "Internal code refactoring"
- ATQTWAA (answers to questions that weren't answered above):
  - Q: What happens with the old media folder?
    A: It just stays there. The db is not populated from the old cache.
  - Q: What happens if two or more minetest instances load media at the same time?
    A: In the worst case, the one that comes later silently goes back to main menu (I hope at least, that there's no worse case). There are some warnings and error messages in the terminal and debug.txt about the database being locked.
  I've tested this by adding a sleep of 10 seconds in the part of the code where the sql transaction is being done.
  - Q: Do you want performance advice? (This is also a NWBAQ (never would be asked question).)
    A: Yes, always.
 - Some more notes about performance:
   * In mtg, looking up the media files takes very roughly 50 ms, which is nothing compared to the ca. 2 s connection time, of which ca. 0.5 s are media loading.
   * Cache-cleaning is most of the times very cheap (ie. roughly 50 ms). It can however take long if nearly a whole fatly filled cache is cleaned. Vacuuming also takes *very* long if there are many files (around 20 s IIRC). This is compensated via multi-threading. The player can still observe this though through a higher cpu load and he has to wait if he quits directly after joining the game.
- Things open to decide:
  * It was suggested on irc to use an expiry logic that keeps a maximum cache size and removes oldest files when the size exceeds the max size.
    - Possible upsides:
      * There's a maximum size.
      * There's an upper limit for the size of files deleted from cache at once: The size of the new files (plus the size of the largest deleted file).
    - Possible downsides:
      * Very old media is kept even if it's never used. => cache needs more disk space
      * If max size is too small, or/and the player has much big media, the cache won't work effectively.
  * Does vacuuming really make sense? It's expensive and Idk how much performance would degrade if we didn't do it.
- A walk through the src code changes:
  * Utility fixes and upgrades:
    - `util/basic_macros.h`: Added `DISABLE_CLASS_MOVE`, similar to `DISABLE_CLASS_COPY`.
    - `threading/thread.{h, cpp}`: Fixed `Thread::start()` blocking indefinitely if the thread's function finishes too fast. (Needed because of a new unittest.)
    - `client.h`, partly `client.cpp`: Added thread vector for async jobs. (Currently only used for one thread ever.)
    - `util/thread.h`: Added `LambdaThread` class for making a `Thread` from a lambda.
    - `util/time_parsing.{h, cpp}`: Added helper for parsing time values, ie. "1.5 ms". (Used for parsing setting values.)
    - `util/Optional.h`: Fixed some problems with `Optional`. Movement semantics should now work. It should match `std::optional` from the too new C++ stdlib. (Needed because I think I needed move semantics somewhere.)
    - `unittest/test_utilities.cpp`: Unittests for most of the above.
    - small parts of `database/database-sqlite3.h`: Added conversion functions for blob data.
  * Main changes:
    - `builtin/settingtypes.txt`, `defaultsettings.cpp`: New settings for expiry heuristics and co..
    - partly `client/client.cpp`, `client/clientmedia.{h, cpp}`: Switched to the cache db, new functions for cache cleaning and co.. There is also some benchmarking code, TODO: remove.
    - `database/database-sqlite3.{h, cpp}`: Added `MediaCacheDatabaseSQLite3` class for the media db. This is where most things happen.
  * Files not mentioned above:
    - `{util, client}/CMakeLists.txt`
    - `client/filecache.{h, cpp}`: Deleted because it's now unused.

## To do

This PR is Ready for deciding whether it suits into the roadmap.

TODO:
* Add an interface for the media db, to allow implementation of other backends.
* Add setting for vacuuming.
* Though there is some code to measure speed **that needs to be removed _before merge_** (the code needs to be removed, not the speed).

## How to test

- First move your old media cache folder, just to be sure.
- Here's a small lua script to insert the old files into the db (db must exist already): https://gist.github.com/Desour/98229b8604c2957ebd4d2791e539bc48
- Start the game. You should notice no new big delay.
- Play around with the new settings. You can look at the db via the cli `sqlite3` tool.
- Read the code.
  There are some (at least one) code paths that are never taken, ie. the decompression which is currently unused. Also, this PR is not the easiest to test by using, so reading the code is important.
- Try to break something.
- Run the unittests (only for a small new utility helper).
